subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -35,13 +35,33 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+/**
+ * Updater for hierarchical file watchers.
+ *
+ * This is the lifecycle for the watched project root directories:
```

Comment:
It's great that we are adding this detailed doc here.

I'd start with why this is happening. Something like this:

```suggestion
 * We want to keep track of project root directories for hierarchical watchers, because ... The project root directories are discovered as included builds are encountered at the start of a build, and then they are removed when the build finishes.
 *
 * This is the lifecycle for the watched project root directories:
 * 
```